src/interfaces/init.h

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_INTERFACES_INIT_H
+#define BITCOIN_INTERFACES_INIT_H
+
+#include <fs.h>
+#include <functional>
+#include <memory>
+#include <string>
+#include <vector>
+
+struct NodeContext;
+
+namespace interfaces {
+class Base;
+class Echo;
+class IpcProcess;
+class IpcProtocol;
+
+//! Init interface providing access to other interfaces. The Init interface is
+//! the interface first exposed when a process spawns or connects to another
+//! process.
+//!
+//! The interfaces::Init interface is similar to other cross-process C++
+//! interfaces like interfaces::Node, interfaces::Wallet, interfaces::Chain and
+//! interfaces::ChainClient, providing virtual methods that can be called from
+//! other processes. What makes it special is that unlike other interfaces which
+//! are not implemented by every process—interfaces::Node is only implemented by
+//! the node process and interfaces::Wallet is only implemented by the wallet
+//! process—interfaces::Init is implemented by every process that supports being
+//! spawned, and it is the initial interface returned by the
+//! IpcProtocol::connect(fd) method, allowing the parent process to control the
+//! child process after the connection is established. The interfaces::Init
+//! interface has methods that allow the parent process to get access to every
+//! interface supported by the child process, and when the parent process frees
+//! the interfaces::Init object, the child process shuts down.
+//!
+//! When spawning a new process, the steps are:
+//!
+//! 1. Client process calls IpcProcess::spawn(), which spawns a new process and
+//!    returns a socketpair file descriptor for communicating with it.
+//! 2. Client process calls IpcProtocol::connect() passing the socketpair
+//!    descriptor, which returns a client Init proxy object calling remote Init
+//!    interface methods.
+//! 3. Client process calls client Init proxy object methods to make new client
+//!    proxy objects calling other remote interfaces. It can also destroy the
+//!    Init object to shut down the spawned process.
+//! 4. Spawned process calls IpcProcess::serve(), to read command line arguments
+//!    and determine that it is a spawned process and what socketpair file
+//!    descriptor it should use. The spawned process then calls
+//!    IpcProtocol::serve() to handle incoming requests from the socketpair and
+//!    invoke Init interface methods and eventually shut down and destroy the
+//!    Init interface when the connection is closed.
```

Comment:
I think an in-depth walk-through needs to be documented somewhere for actual and future reviewers.

An example would be :
1. `Bitcoin GUI process is started, in `src/qt/bitcoin.cpp`, it initializes its own interface `LocalInit` from `src/interfaces/init_bitcoin-gui.cpp`
2. From this `LocalInit` interface, it calls `makeNode()`, which trigger a child `bitcoin-node` spawning through `IpcProcess.spawn`. It also register hook cleaning through `base.addCloseHook`
3. From this `LocalInit` interface, it calls `connect`, which start a new thread `capnp-loop` with a libmultiprocess `mp::EventLoop`
4. Bitcoin Node is started with `-ipcfd` passing a socketpair for communications, it initializes its own interface `LocalInit` from `src/interfaces/init_bitcoin-node.cpp`
5. From this `LocalInit` interface, it calls `serve()` to start servicing Bitcoin GUI on its owned interface `Chain`. This is achieved through the abstract interface `IpcProtocol` which is currently Cap'n Proto (`MakeCapnpProtocol`)
6. When Bitcoin GUI request a `Chain` method like `hasBlocks`, this is redirected to `mp::ProxyClient`
7. From its own `capnp-loop` thread embedding a `mp::ProxyServer`, Bitcoin Node internal methods are called to fulfill the request

Okay it's likely a fuzzy walkthrough but this reflects my understanding right now.

Reply:
Thanks for the new documentation in `multiprocess.md`.

For the text representation I was thinking a simple ASCII art underlying relations between interfaces, implementations, process and threads. Akin to RL's [ARCH.md](https://github.com/rust-bitcoin/rust-lightning/blob/main/ARCH.md), which is hopefully straightforward and doesn't rely on a third-party tool.

I'm working on a proposal, if it will be easier to give your opinion.